[UPDATED] TCP Write and SlowConsumer handling

kozlovic opened the pull request:
- All writes will now be done by the writeLoop, unless when the
  writeLoop has not been started yet (likely in connection init).
- Slow consumers for non CLIENT connections will be reported but
  not failed. The idea is that routes, gateway, etc.. connections
  should stay connected as much as possible. However if a flush
  operation times out and no data at all has been written, the
  connection will be closed (regardless of type).
- Slow consumers due to max pending is only for CLIENT connections.
  This allows sending of SUBs through routes, etc.. to not have
  to be chunked.
- The backpressure to CLIENT connections is increased (up to 1sec)
  based on the sub's connection pending bytes level.
- Connection is flushed on close from the writeLoop as to not block
  the "fast path".

Some tests have been fixed and adapted since now closeConnection()
is not flushing/closing/removing connection in place.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

derekcollison commented on server/client.go:
```diff
@@ -1116,6 +1102,90 @@ func (c *client) flushOutbound() bool {
+		c.out.wdl, numChunks, attempted)
+
+	// We always close CLIENT connections, or when nothing was written at all...
+	if c.kind == CLIENT || written == 0 {
```
Can written == 0 be legit though on some platforms?

kozlovic replied:
Yes, but then we would never "slow consumer" non client connections. The written == 0 follows a write deadline error, so it means that nothing was written for the write deadline. I think it is reasonable in this case to kill the connection.

derekcollison replied:
Ah sorry did not pick up that is was post write deadline error.

derekcollison commented on server/client.go:
```diff
@@ -1116,6 +1102,90 @@ func (c *client) flushOutbound() bool {
+	// Save off the connection if its a client or leafnode.
+	if c.kind == CLIENT || c.kind == LEAF {
+		if nc := c.nc; nc != nil && c.srv != nil {
+			go c.srv.saveClosedClient(c, nc, reason)
```
This works, but I think longer term we need to do something different TBH. Maybe just put a TODO comment here.

kozlovic replied:
Not sure what you are referring to? Meaning, I can put a TODO but about what?

kozlovic replied:
Not sure what TODO you want me to put here? Not sure I understand what's not working long term..

derekcollison replied:
Each save uses a Go routine. We may want to do something different longer term and send events to a single Go routine.

kozlovic replied:
Done

derekcollison commented on server/client.go:
```diff
@@ -1116,6 +1102,90 @@ func (c *client) flushOutbound() bool {
+func (c *client) expectConnect() bool {
+	switch c.kind {
+	case CLIENT:
+		return true
```
This is not always true. Think telnet example.

kozlovic replied:
The problem is that we treat a write timeout differently when the CONNECT has not been received yet. That is, there is a test that make sure that if the client does not send the CONNECT and the write deadline is hit, we do not report has a slow consumer (will be auth timeout instead).

derekcollison replied:
If server has auth set then I think it's ok to say client needs to send connect, but when its not sent I think this should return false for client. WDYT?

kozlovic replied:
Ok, this was more involved since I had to access server's info's AuthRequired or TLSRequired, so I moved to flags instead and set it in each of the createClient(), createRoute(), etc... with the same logic that was in expectConnect() switch case.

derekcollison replied:
TLS for clients does not mean connect is needed, only auth IIRC.

derekcollison commented on server/client.go:
```diff
@@ -1536,20 +1606,33 @@ func (c *client) queueOutbound(data []byte) bool {
 // Assume the lock is held upon entry.
-func (c *client) sendProto(info []byte, doFlush bool) {
-	if c.nc == nil {
+func (c *client) enqueueProtoAndFlush(proto []byte, doFlush bool) {
```
I get the change but.. ;)

I think sendProto is ok here with comments about what it does.

derekcollison replied:
But since you have higher level functions that look a bit nicer probably ok.

kozlovic replied:
The goal really was to avoid issues that if you made sendProto() calls in the JS branch, I want you to have to resolve the conflict and understand that the new call does now. Wanted also to make sure that the memory muscle of using sendProto() in the future fails :-)

derekcollison replied:
Good point, and agree now that I reviewed it all.

derekcollison commented on server/client.go:
```diff
@@ -1536,20 +1606,33 @@ func (c *client) queueOutbound(data []byte) bool {
 func (c *client) sendPong() {
 	c.traceOutOp("PONG", nil)
-	c.sendProto([]byte("PONG\r\n"), true)
+	c.enqueueProto([]byte("PONG\r\n"))
```
Should make a const?

kozlovic replied:
Will do..

derekcollison commented on server/client.go:
```diff
@@ -1583,7 +1666,7 @@ func (c *client) sendPing() {
 	c.ping.out++
 	c.traceOutOp("PING", nil)
-	c.sendProto([]byte("PING\r\n"), true)
+	c.enqueueProto([]byte("PING\r\n"))
```
Same here for const if possible.

kozlovic replied:
Done

derekcollison commented on server/client.go:
```diff
@@ -1598,32 +1681,26 @@ func (c *client) generateClientInfoJSON(info Info) []byte {
 }
 
 func (c *client) sendOK() {
 	proto := []byte("+OK\r\n")
```
Same here w/ const.

kozlovic replied:
Done

derekcollison commented on server/client.go:
```diff
@@ -1807,7 +1884,8 @@ func (c *client) processSub(argo []byte, noForward bool) (*subscription, error)
 
-	if c.nc == nil && kind != SYSTEM {
+	// This check does not apply to SYSTEM clients (because they don't have a `nc`...)
+	if c.isClosed() && kind != SYSTEM {
```
Would reverse them. 

```go
if kind != SYSTEM && c.isClosed()
```

kozlovic replied:
Done

derekcollison commented on server/client.go:
```diff
@@ -2249,13 +2327,20 @@ func (c *client) msgHeader(mh []byte, sub *subscription, reply []byte) []byte {
 	stall := c.out.stc
+	ttl := 100 * time.Millisecond
+	if c.out.pb >= c.out.mp {
+		ttl = time.Second
```
That feels like a long time..

kozlovic replied:
Since we don't "slow consumer" the routes/gws/ln, I wanted to put enough back pressure on the clients to avoid memory build-up. We can always revisit, but I think for now it is ok.

derekcollison replied:
Agree regarding revisit. I am working in JetStream branch on limits of memory etc and have a decent way I do that, we can do something similar here where we have a spike memory budget that once exceeded we will stall and release when it drops below.

derekcollison commented on server/client.go:
```diff
@@ -2363,7 +2448,7 @@ func (c *client) deliverMsg(sub *subscription, subject, mh, msg []byte, gwrply b
 
 	// Check for closed connection
-	if client.flags.isSet(clearConnection) {
+	if client.flags.isSet(closeConnection) {
```
```go
client.isClosed()
```

Since you use that one mostly.

kozlovic replied:
Done

derekcollison commented on server/client.go:
```diff
@@ -3188,8 +3273,9 @@ func (c *client) processPingTimer() {
-			c.sendProto([]byte(fmt.Sprintf("-ERR '%s'\r\n", "Stale Connection")), true)
-			c.clearConnection(StaleConnection)
+			c.enqueueProto([]byte(fmt.Sprintf("-ERR '%s'\r\n", "Stale Connection")))
+			c.mu.Unlock()
```
Did this dangle holding the lock forever before??

kozlovic replied:
No, just that I need to call c.closeConnection() now and so need to release the lock (used to be a lock/defer unlock).

derekcollison replied:
Still not following, I just saw the unpaired unlock so was confused. Do you re-acquire?

kozlovic replied:
Top used to be lock/defer unlock and here we used to have c.sendProto()/c.clearConnection() and return. I am now explicit with lock/unlock because I need to call closeConnection() (line below this one) and that requires to be done outside of lock.

derekcollison replied:
ok thanks for clarifications.

derekcollison commented on server/leafnode.go:
```diff
@@ -1092,7 +1099,7 @@ func (c *client) updateSmap(sub *subscription, delta int32) {
 		delete(c.leaf.smap, key)
 	}
-	if update && c.flags.isSet(leafAllSubsSent) {
+	if update {
```
This change was for a specific condition, not applicable anymore? Was a very specific bug. I think we have a test for it.

kozlovic replied:
No because we no longer split/flush in place/release lock/etc..

derekcollison replied:
Ah, gothca.

kozlovic commented:
@derekcollison Pushed a separate commit (https://github.com/nats-io/nats-server/pull/1233/commits/fac6b7328bb642616589691c9f2f7f5b2dd1978c) with the changes to ease the review. I think I have addressed all comments. I will leave to 1sec the stall for now, can revisit later (see my comment).

kozlovic commented:
humm... the new test with the CONNECT/Slow consumer seem to be failing on Travis (works fine on Mac). Will have to tweak it..

kozlovic commented:
Ok, all good now.

derekcollison commented on server/server.go:
```diff
@@ -1673,6 +1673,9 @@ func (s *Server) createClient(conn net.Conn) *client {
 
 	// Grab lock
 	c.mu.Lock()
+	if info.AuthRequired || info.TLSRequired {
```
TLSRequired != connect required IIRC.
